### ISSUE TITLE
Swagger UI & Token Based auth

### ISSUE THREAD

**Issue by wkochman** [jhipster/generator-jhipster#1697]
The swagger UI is not working with token based auth. 
The token is not passed to the request.

**Comment by jdubois** [comment]
Yes, this is correct -> this isn't a new issue, in fact, as we already had the issue with Swagger 1, so migrating to SpringFox didn't create a new issue
But I forgot to add this to the ticket system, my bad!

**Comment by cbornet** [comment]
Swagger ui doesn't support the oauth2 password flow which is used by jhipster
See https://github.com/swagger-api/swagger-ui/issues/807
Only auth code and implicit grants are supported.

**Comment by gmarziou** [comment]
A potential solution for this would be to copy the springfox HTML page into JHipster webapp so that we can add some code to read the token from localstorage and attach it to iframe window this is [how it was done previously in JHIpster](https://github.com/jhipster/generator-jhipster/commit/d5a94adc95a20fe8c7b58c142301d7b434673f89).

Most of the problems come from the fact that swagger-ui is a JS application on its own and that the only way to integrate it is to use an iframe, so each time there 's a different authentication scheme, either we wait for swagger-ui to support it as @cbornet reported it for oauth2 password flow or we have to copy their HTML and modify it (so basically fork it with all drawbacks it implies).

So, let me summarize potential solutions:

- we request swagger-ui to provide a hook to let applications integrate their authentication/authorization requirements
- we stop using swagger-ui page and turn to something we can intregrate more easily like an angularjs directive. I found 2 such projects [angular-swagger](https://github.com/dnmilne/angular-swagger) and [angular-swagger-ui](https://github.com/Orange-OpenSource/angular-swagger-ui).
- we explain JHipster users that swagger-ui may not work with some auth schemes unless they decide to make it public.

**Comment by dilipkrish** [comment]
While any of the suggested solutions like angular-swagger and angular-swagger-ui are certainly viable. However, think there is some level of customization/configurationn that needs to provided via JHipster regardless. I personally prefer the springfox approach since its driven by spring configurations. 

We already support to some extent [bootstrapping the security configuration](http://springfox.github.io/springfox/docs/current/#springfox-swagger2-with-spring-boot) see # 19. Rather than use localStorage/cookies etc. its just a bean that is served by the provided **_springfox meta-services_**

```java
@Bean
  SecurityConfiguration security() {
    return new SecurityConfiguration(
        "test-app-client-id",
        "test-app-realm",
        "test-app",
        "apiKey");
  }
```

So in order to support the password flow oauth scheme and any additional schemes, we need to do a few things.
- Add support for the password flow oauth scheme in swagger ui (the swagger team is pretty receptive to PRs and ideas)
- Augment the `SecurityConfiguration` __if required__ to support password flow
- Modify the [swagger-ui.html](https://github.com/springfox/springfox/blob/master/springfox-swagger-ui/src/web/swagger-ui.html) which is a modified index.html from swagger-ui, and [springfox.js](https://github.com/springfox/springfox/blob/master/springfox-swagger-ui/src/web/js/springfox.js) to support password flow.

One caveat is that I haven't been able to load [these beans](https://github.com/springfox/springfox/blob/master/springfox-swagger-common/src/main/java/springfox/documentation/swagger/web/ApiResourceController.java#L52-L55) successfully using jhipster, and springfox **2.1.2** so will defenitely need help with that. Since those beans are not required, the controller is eagerly loaded into the spring context and by the time the bean is defined in the application its too late.

I'm happy to help out anyway I can cheers! :beers:

**Comment by gmarziou** [comment]
Thanks a lot for your responsiveness.

I did not get that we were using springfox's swagger-ui.html.

I din't underts

There's one thing that I  don't understand, what you presented is mainly server-side configuration exposed to client and read by springfox.js. For oauth2 authentication or token-based

**Comment by gmarziou** [comment]
Thanks a lot for your responsiveness.

I did not get that we were using springfox's swagger-ui.html.

If I understand correctly the SecurityConfiguration server-side bean would tell client-side springfox.js that this is a token based security and where to find the token. So for instance it could be a key and the storage service like 'window.localStorage'.

**Comment by cbornet** [comment]
Also I don't see the advantage of using password flow Oauth2 alone compared to plain old basic auth... Especially in JS applications where you can't even hide the client_secret.

**Comment by dilipkrish** [comment]
> If I understand correctly, the SecurityConfiguration server-side bean would tell client-side springfox.js that this is a token based security and where to find the token. So for instance it could be a key and the storage service like LocalStorage or SessionStorage or even a cookie.

Not really. I don't claim to be an expert in oauth, but her is my understanding
- The oauth protocol and flow handles how the token exchange dance happens. So I don't think we need to specify the client side storage and retrieval model etc.
- In addition  apiKey and basic auth, swagger-ui exposes a model to configure the different players in this oauth dance as well, i.e. login url, token endpoint url, oauth flow etc. This is the piece we're leveraging using the spring beans.

Also like @cbornet suggests, there isn't much value in using oath2 password profile over basic auth.

**Comment by gmarziou** [comment]
But JHipster supports also x-auth-token authentication, so in this case my proposition would be correct I guess.

**Comment by gmarziou** [comment]
I think the value in using oauth2 password profile over basic auth is that the credentials are sent only once to the server rather than with each request and also there's a logout mechanism.

**Comment by dilipkrish** [comment]
👍🏻 

In any case, I do need some help to resolve the bean resolution problem I mentioned above 👆🏻

**Comment by gmarziou** [comment]
Do you have some code to share?

**Comment by dilipkrish** [comment]
I have a fork with [just a simple change](https://github.com/dilipkrish/jhipster-sample-app/commit/88947110bfc546b3ffe17692d47e6976794e76b0). Upgraded the dependencies locally to 2.1.2 and added the security bean.

I was expecting [this bean](https://github.com/springfox/springfox/blob/master/springfox-swagger-common/src/main/java/springfox/documentation/swagger/web/ApiResourceController.java#L52-L53) to be resolved to the provided bean. and when I hit `http://localhost:8080/configuration/security` I'd expect it to return.

```json
{
    "clientId":"123",
    "realm":"abc",
    "appName":"test",
    "apiKey":"123456"
}
```

**Comment by gmarziou** [comment]
@dilipkrish I got it working just by renaming the bean method in SwaggerConfiguration because I think it was conflicting with JHipster own SecurityConfiguration.

~~~~
    @Bean
    public SecurityConfiguration sfoxSecurityConfiguration() {
        return new SecurityConfiguration("123", "abc", "test", "123456");
    }
~~~~

**Comment by dilipkrish** [comment]
Doh!!! Thanks for solving the mystery :) 

That should've also demonstrated what I meant in the discussion above. Basically exposing a configuration in combination with the specified security schemes in the `Docket`

**Comment by cbornet** [comment]
@gmarziou With basic auth, you generally use the session to provide the login/logout/expire/don't send credentials everytime that is in Oauth2.
In fact the power of Oauth2 is authorization, not authentication. Oauth2 in JHipster would truly be interesting if we implemented the authorization code flow with an authorization form to do 3-legged auth of the API. Otherwise, I would stick to basic auth.

**Comment by gmarziou** [comment]
@cbornet I misunderstood you, I thought you were talking about http basic auth while you were talking of jhipster session auth.

**Comment by gmarziou** [comment]
@dilipkrish I don't understand how your proposal works, so I will try to re-implement our logic in our own copy of swagger-ui/index.html which retrieves token from localstorage to fixthis issue and related ones in short term then I'll see what can be done for longer term.

**Comment by cbornet** [comment]
@gmarziou Yes, my first comment was not very accurate, sorry... :-)

**Comment by dilipkrish** [comment]
@gmarziou youre better qualified so I'll leave it in your capable hands. If there is something I can pull in to `springfox-swagger-ui`  to make the packaging easier let me know.

**Comment by gmarziou** [comment]
Thanks, i'll probably have a better understanding after this first step.
